### PR TITLE
feat: fix abtestui overlay

### PR DESCRIPTION
**Pull request by kamilkifer**
</br>

### What ⁉️
- fix the AB test ui.
  - problem caused by the app launching BEFORE the user config was loaded, leading to it being initialised with default config values.

### Why ⁉️
- To make it slightly easier to preview various variations of A/B tests.

### How ⁉️
- npm link skift into wrapper
- npm link wrapper into relevant app
- slowly but surely, using the debugger.

- The issue was that the app was running the shouldShowUI() call, before the user config was passed in.
- currently handled by the configLoaded variable.

### How to review and test 📱
- npm link as mentioned above.

### Extra info
- More problems spawned from the GHA
  - Initially, the linter was complaining, and the cause was `import type`
  - After some research, type-only imports were introduced in [TS 3.8](https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html)
- Upgrading caused further compatibility problems.
- I've updated a few packages, which causes other compatibility errors to appear, and lead to more package upgrades.
- Eventually got it all working locally (via npm link), with tests passing and building properly.

</br>

### Checklist ✅
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing tests are passing locally

### CODE REVIEW

File: tests/api.spec.ts [hunk 1]
```diff
@@ -1,3 +1,7 @@
+/**
+ * @jest-environment jsdom
+ */
+
```
Comment (kamilkifer) [inline tests/api.spec.ts:4]: fixes `ReferenceError: navigator is not defined`

File: package.json [hunk 2]
```diff
@@ -35,25 +35,26 @@
     "@semantic-release/changelog": "^6.0.1",
     "@semantic-release/git": "^10.0.1",
     "@types/jest": "^24.0.18",
-    "@types/node": "^12.7.8",
+    "@types/node": "^20.12.12",
```
Comment (fernicolosi) [inline package.json:38]: I can see that you're using v16 in the `.nvmrc` file, any reason to have different node version here? Maybe we can try to move to v20. You'll want to re-generate the package lock file

File: .github/workflows/release.yml [hunk 3]
```diff
@@ -11,7 +11,7 @@ jobs:
     steps:
       - uses: actions/checkout@v3
       - uses: actions/setup-node@v3
-      - run: npm ci
+      - run: npm ci --legacy-peer-deps
```
Comment (fernicolosi) [inline .github/workflows/release.yml:14]: Can we avoid this flag? Is there some conflicting version of dependencies that this is trying to address?

File: src/ui.dev.ts [hunk 4]
```diff
@@ -57,14 +56,14 @@ export const uiFactory = (
     ) {
         const item = document.createElement('li');
         item.textContent = variation.name;
-        const open = renderButton(require('./images/open.svg'), () => {
+        const open = renderButton('change to this variant', () => {
             setCurrentTestVariation(splitTest.name, variation.name);
         });
 
-        const link = renderLink(splitTest, variation);
+        // const link = renderLink(splitTest, variation);
```
Comment (fernicolosi) [inline src/ui.dev.ts:63]: Can we remove these commented out lines?

File: src/index.ts [hunk 5]
```diff
@@ -13,7 +13,8 @@ import {
 } from './main';
 import { SplitTest } from './splitTest';
 import { uiFactory } from './ui';
-import { CookiePersister, UserSessionPersister } from './userSessionPersister';
```
Comment (fernicolosi) [inline src/index.ts:16]: What's the reason for this change in imports?

Comment (kamilkifer) [reply src/index.ts:16]: Hey, 
TS was a bit outdated and was having trouble with the Type import.
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-3-8.html#type-only-imports-and-export

Comment (fernicolosi) [reply src/index.ts:16]: Nice, thanks!

File: .github/workflows/release.yml (hunk 3 above)
Comment (kamilkifer) [reply .github/workflows/release.yml]: Yes, there was a conflict with the css-to-string-loader - I've found an alternative loader, which works and loads the css properly (at least via npm link)
[Updated to v20 here](https://github.com/trustpilot/skift/pull/126/commits/1a4a33348f71e7562aa0f278a349868fdf1213a0)
[Removed the flag here](https://github.com/trustpilot/skift/pull/126/commits/b02ac4d50f52914c5aaca7cd5e9ff419fd8ef725)

File: package.json (hunk 2 above)
Comment (kamilkifer) [reply package.json:38]: [v20 here](https://github.com/trustpilot/skift/pull/126/commits/1a4a33348f71e7562aa0f278a349868fdf1213a0)

File: src/index.ts (hunk 5 above)
Comment (kamilkifer) [reply src/index.ts:16]: Update:
I believe this will have to stay for now - after running locally I keep running into this error when `import type` isn't present.
export 'UserSessionPersister' (reexported as 'UserSessionPersister') was not found in './userSessionPersister' (possible exports: CookiePersister)

File: src/ui.dev.ts (hunk 4 above)
Comment (kamilkifer) [reply src/ui.dev.ts]: [Removed all here.](https://github.com/trustpilot/skift/pull/126/commits/1a4a33348f71e7562aa0f278a349868fdf1213a0#diff-164d12e0f8d306810bbb6b672e8115434dc03fc6a9fc245292bac3558b0d537c)

File: src/main.ts [hunk 6]
```diff
@@ -8,6 +8,7 @@ import _getUserAgentInfo from './userAgentInfo';
 import userSession, { UserSession } from './userSession';
 
 const userAgentInfo = _getUserAgentInfo();
+let configLoaded = false;
```
Comment (fernicolosi) [inline src/main.ts:11]: I wonder, wouldn't it make sense to make this a prop of the `_config` option?

File: src/main.ts [hunk 7]
```diff
@@ -123,7 +128,20 @@ export function reset(): void {
...
+                return;
+            }
+            clearInterval(interval);
+            resolve();
+        }, checkInterval);
+    });
+};
+
 export async function shouldShowUI() {
+    await waitUntil(() => configLoaded);
```
Comment (fernicolosi) [inline src/main.ts:144]: Could it be a problem if `configLoaded` is not set to `true`? Could that even happen? 🤔

File: src/ui.dev.ts [hunk 8]
```diff
@@ -27,13 +27,12 @@ export const uiFactory = (
     }
 
     function renderLink(splitTest: SplitTest, variation: InternalVariation) {
-        return renderButton(require('./images/link.svg'), () => {
-            const input = document.createElement('input');
-            input.value = splitTest.getVariationUrl(variation.name);
-            document.body.appendChild(input);
-            input.select();
+        return renderButton('currently active', () => {
```
Comment (fernicolosi) [inline src/ui.dev.ts:30]: Haven't run it locally, but any reason for not showing the .svg anymore?

File: tests/api.spec.ts (hunk 1 above)
Comment (fernicolosi) [reply tests/api.spec.ts:4]: Looks good, curious if you consider adding a `jest.config` file to set the environment across all tests

File: src/ui.dev.ts (hunk 8 above)
Comment (kamilkifer) [reply src/ui.dev.ts:30]: It looked a bit nasty before:
image


Now:
image



Not ideal, but probably a slightly better option this way (subjective).

Comment (fernicolosi) [reply src/ui.dev.ts:30]: oh, yeah. I think that looks better! Do we need to clean up that .svg or is it used elsewhere?

File: src/main.ts (hunk 6 above)
Comment (kamilkifer) [reply src/main.ts]: It would!
I did a similar thing [HERE](https://github.com/trustpilot/skift/blob/7f082b076f7a4f278cb3b15b06fc179118d9deb1/src/main.ts#L155-L156), but instead used the cookieName from the splitTest package.
The default config in skift uses "skiftABtest" and I'm now waiting for the proper cookieName before carrying on with the shouldShowUI() function.

Also auto-resolves the promise after 2 seconds to avoid the promise hanging forever.
(Keep in mind this behaviour is only in staging, [ui prod](https://github.com/trustpilot/skift/blob/7f082b076f7a4f278cb3b15b06fc179118d9deb1/src/ui.prod.ts#L5) does not use the shouldShowUI() function.)

File: src/ui.dev.ts (hunk 8 above)
Comment (kamilkifer) [reply src/ui.dev.ts:30]: Yep, missed that!
[removed here.](https://github.com/trustpilot/skift/pull/126/commits/febad263268bcfd851d4a6d579a1e09a60ecbe7c)

File: tests/query.spec.ts [hunk 9]
```diff
@@ -25,6 +25,9 @@ describe('Query', () => {
 
     describe('#getAbTestParameter', () => {
         it('should return null if there is no query', () => {
+            const location = {
```
[the remark this answers is not in the document]
Comment (kamilkifer) [reply tests/query.spec.ts]: No, removed now.
Sorry got sloppy since i've not looked at it for a while.

[the remark this answers is not in the document]
Comment (fernicolosi) [reply tests/query.spec.ts]: No worries 👍🏼

**Review by fernicolosi — APPROVED** [review]
